File: AdPieSample/AdPieSample/Source/InterstitialAdViewController.m
#import "InterstitialAdViewController.h"
#import <AdPieSDK/AdPieSDK.h>
#import "UIViewController+Toast.h"

@interface InterstitialAdViewController () <APInterstitialDelegate>
// Private Properties
@property (nonatomic, copy) NSString *slotId;
@property (nonatomic, strong) APInterstitial *interstitial;
@property (nonatomic, strong) UIButton *loadButton;
@property (nonatomic, strong) UIButton *showButton;
@property (nonatomic, strong) UIStackView *stackView;
@end

@implementation InterstitialAdViewController

#pragma mark - Initialization

- (instancetype)initWithSlotId:(NSString *)slotId {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _slotId = [slotId copy];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    return self;
}

#pragma mark - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    [self setupUI];
    [self setupActions];
}

#pragma mark - UI Setup

- (void)setupUI {
    // 1. Load Button
    self.loadButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.loadButton setTitle:@"Load Ad" forState:UIControlStateNormal];
    self.loadButton.backgroundColor = [UIColor systemBlueColor];
    [self.loadButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.loadButton.layer.cornerRadius = 8;
    self.loadButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightBold];
    self.loadButton.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.loadButton.heightAnchor constraintEqualToConstant:50].active = YES;
    [self.loadButton.widthAnchor constraintEqualToConstant:200].active = YES;
    
    // 2. Show Button
    self.showButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.showButton setTitle:@"Show Ad" forState:UIControlStateNormal];
    self.showButton.backgroundColor = [UIColor systemGrayColor];
    [self.showButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.showButton.layer.cornerRadius = 8;
    self.showButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightBold];
    self.showButton.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.showButton.heightAnchor constraintEqualToConstant:50].active = YES;
    [self.showButton.widthAnchor constraintEqualToConstant:200].active = YES;
    
    // 3. StackView
    self.stackView = [[UIStackView alloc] init];
    self.stackView.axis = UILayoutConstraintAxisVertical;
    self.stackView.spacing = 20;
    self.stackView.alignment = UIStackViewAlignmentCenter;
    self.stackView.distribution = UIStackViewDistributionFill;
    self.stackView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.stackView addArrangedSubview:self.loadButton];
    [self.stackView addArrangedSubview:self.showButton];
    
    [self.view addSubview:self.stackView];
    
    // 4. StackView AutoLayout
    [NSLayoutConstraint activateConstraints:@[
        [self.stackView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.stackView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor]
    ]];
}

- (void)setupActions {
    [self.loadButton addTarget:self action:@selector(didTapLoadAdButton) forControlEvents:UIControlEventTouchUpInside];
    [self.showButton addTarget:self action:@selector(didTapShowAdButton) forControlEvents:UIControlEventTouchUpInside];
}

#pragma mark - Actions

- (void)didTapLoadAdButton {
    self.interstitial = [[APInterstitial alloc] initWithSlotId:self.slotId];
    self.interstitial.delegate = self;
    [self.interstitial load];
}

- (void)didTapShowAdButton {
    if (!self.interstitial) { return; }
    
    if ([self.interstitial isReady]) {
        [self.interstitial presentFromRootViewController:self];
    }
}

#pragma mark - APInterstitialDelegate

- (void)interstitialDidLoadAd:(APInterstitial *)interstitial {
    [self showToastWithMessage:[NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__]];
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialDidFailToLoadAd:(APInterstitial *)interstitial
                          withError:(NSError *)error
{
    [self showToastWithMessage:[NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__]];
    NSString * errorMsg = [NSString stringWithFormat:@"Error (code : %d, message : %@, date : %@)",
                           (int)[error code],
                           [error localizedDescription],
                           [[NSDate date] descriptionWithLocale:[NSLocale systemLocale]]];
    NSLog(@"%s, %@", __PRETTY_FUNCTION__, errorMsg);
}

- (void)interstitialDidFailToShowAd:(APInterstitial *)interstitial withError:(NSError *)error {
    [self showToastWithMessage:[NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__]];
    NSString * errorMsg = [NSString stringWithFormat:@"Error (code : %d, message : %@, date : %@)",
                           (int)[error code],
                           [error localizedDescription],
                           [[NSDate date] descriptionWithLocale:[NSLocale systemLocale]]];
    NSLog(@"%s, %@", __PRETTY_FUNCTION__, errorMsg);
}

- (void)interstitialDidDismissScreen:(APInterstitial *)interstitial {
    [self showToastWithMessage:[NSString stringWithFormat:@"%s",__PRETTY_FUNCTION__]];
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialWillPresentScreen:(APInterstitial *)interstitial {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialWillDismissScreen:(APInterstitial *)interstitial {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialWillLeaveApplication:(APInterstitial *)interstitial {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

@end

File: AdPieSwiftSample/AdPieSwiftSample/Source/InterstitialAdViewController.swift
import UIKit
import AdPieSDK

class InterstitialAdViewController: UIViewController {

    let slotId: String
    var interstitial: APInterstitial?
    
    init(slotId: String) {
        self.slotId = slotId
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private let loadButton: UIButton = {
        let button = UIButton(type: .system)
        button.setTitle("Load Ad", for: .normal)
        button.backgroundColor = .systemBlue
        button.setTitleColor(.white, for: .normal)
        button.layer.cornerRadius = 8
        button.titleLabel?.font = UIFont.systemFont(ofSize: 16, weight: .bold)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.heightAnchor.constraint(equalToConstant: 50).isActive = true
        button.widthAnchor.constraint(equalToConstant: 200).isActive = true
        return button
    }()

    private let showButton: UIButton = {
        let button = UIButton(type: .system)
        button.setTitle("Show Ad", for: .normal)
        button.backgroundColor = .systemGray
        button.setTitleColor(.white, for: .normal)
        button.layer.cornerRadius = 8
        button.titleLabel?.font = UIFont.systemFont(ofSize: 16, weight: .bold)
        
        button.translatesAutoresizingMaskIntoConstraints = false
        button.heightAnchor.constraint(equalToConstant: 50).isActive = true
        button.widthAnchor.constraint(equalToConstant: 200).isActive = true
        return button
    }()

    private let stackView: UIStackView = {
        let stack = UIStackView()
        stack.axis = .vertical
        stack.spacing = 20
        stack.alignment = .center
        stack.distribution = .fill
        stack.translatesAutoresizingMaskIntoConstraints = false
        return stack
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        setupUI()
        setupActions()
    }

    private func setupUI() {
        stackView.addArrangedSubview(loadButton)
        stackView.addArrangedSubview(showButton)
        view.addSubview(stackView)
        NSLayoutConstraint.activate([
            stackView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            stackView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
        ])
    }
    
    private func setupActions() {
        loadButton.addTarget(self, action: #selector(didTapLoadAdButton), for: .touchUpInside)
        showButton.addTarget(self, action: #selector(didTapShowAdButton), for: .touchUpInside)
    }
    
    @objc private func didTapLoadAdButton() {
        self.interstitial = APInterstitial(slotId: self.slotId)
        self.interstitial?.delegate = self
        self.interstitial?.load()
    }
    
    @objc private func didTapShowAdButton() {
        guard let interstitial = self.interstitial else { return }
        guard interstitial.isReady() == true else { return }
        interstitial.present(fromRootViewController: self)
    }
}

extension InterstitialAdViewController: APInterstitialDelegate {
    
    // MARK: - APInterstitial delegates
    
    func interstitialDidLoadAd(_ interstitial: APInterstitial!) {
        showToast(message: #function)
        print(#function)
    }
    
    func interstitialDidFail(toLoadAd interstitial: APInterstitial!, withError error: Error!) {
        showToast(message: #function)
        let errorMessage = "Failed to load interstitial ads." + "(code : " + String(error._code) + ", message : " + error.localizedDescription + ")"
        let alertController = UIAlertController(title: "Error", message: errorMessage, preferredStyle: .alert)
        let defaultAction = UIAlertAction(title: "OK", style: .default, handler: nil)
        alertController.addAction(defaultAction)
        present(alertController, animated: true, completion: nil)
    }
    
    func interstitialDidFail(toShowAd interstitial: APInterstitial!, withError error: (any Error)!) {
        showToast(message: #function)
        print(#function)
    }
    
    func interstitialDidDismissScreen(_ interstitial: APInterstitial!) {
        showToast(message: #function)
        print(#function)
    }
    
    func interstitialWillPresentScreen(_ interstitial: APInterstitial!) {
        print(#function)
    }
    
    func interstitialWillDismissScreen(_ interstitial: APInterstitial!) {
        print(#function)
    }
    
    func interstitialWillLeaveApplication(_ interstitial: APInterstitial!) {
        print(#function)
    }
    
}

